File: retort-core/src/main/lombok/co/kaioru/retort/builder/CommandBuilder.java
package co.kaioru.retort.builder;

import co.kaioru.retort.*;
import co.kaioru.retort.exception.CommandException;
import co.kaioru.retort.exception.CommandNotBuiltException;

public class CommandBuilder<I extends ICommandContext, O> extends AbstractCommand<I, O> implements ICommandBuilder<I, O> {

	private ICommandExecutable<I, O> executable;

	public CommandBuilder(String name) {
		super(name);
	}

	@Override
	public ICommandBuilder withDescription(String description) {
		this.setDescription(description);
		return this;
	}

	@Override
	public ICommandBuilder withAlias(String alias) {
		this.getAliases().add(alias);
		return this;
	}

	@Override
	public ICommandBuilder withMiddleware(ICommandMiddleware<I> middleware) {
		this.getMiddlewares().add(middleware);
		return this;
	}

	@Override
	public ICommandBuilder withCommand(ICommand<I, O> command) {
		this.getCommands().add(command);
		return this;
	}

	@Override
	public ICommand<I, O> build(ICommandExecutable<I, O> executable) {
		this.executable = executable;
		return this;
	}

	@Override
	public O execute(I i) throws CommandException {
		if (executable == null)
			throw new CommandNotBuiltException();
		return executable.execute(i);
	}

}

File: retort-core/src/main/kotlin/co/kaioru/retort/builder/impl/CommandBuilder.kt
package co.kaioru.retort.builder.impl

import co.kaioru.retort.ICommand
import co.kaioru.retort.ICommandExecutable
import co.kaioru.retort.builder.ICommandBuilder
import co.kaioru.retort.exceptions.CommandNotBuiltException
import co.kaioru.retort.impl.Command
import co.kaioru.retort.impl.CommandContext

class CommandBuilder<I : CommandContext, O>(name: String) : Command<I, O>(name), ICommandBuilder<I, O> {
    var executable: ICommandExecutable<I, O>? = null

    override fun build(executable: ICommandExecutable<I, O>): ICommand<I, O> {
        this.executable = executable
        return this
    }

    override fun execute(input: I): O {
        if (executable == null)
            throw CommandNotBuiltException()
        return executable?.execute(input)!!
    }
}